Outbox dead letter queue and Observability using Open Telemetry (#1673)
cap-js/docs: java/outbox.md

[Learn more about `EventContext.proceed()`.](./event-handlers/#proceed-on){.learn-more}

## Outbox Dead Letter Queue

The transactional outbox tries to process each entry a specific number of times. The number of attempts is configurable per outbox by setting the configuration `cds.outbox.services.<key>.maxAttempts`.

[Learn more about CDS Properties.](./developing-applications/properties){.learn-more}

Once the maximum number of attempts is exceeded, the corresponding entry is not touched anymore and hence it can be regarded as dead. Dead outbox entries are not deleted automatically, they remain in the database forever. By defining a CDS service the dead entries can be managed. The example below provides a step-by-step guide to develop a Dead Letter Queue for the transactional outbox.

In the first step the service needs to be defined in CDS:

::: code-group

```cds [srv/outbox-dead-letter-queue-service.cds]
using from '@sap/cds/srv/outbox';

service OutboxDeadLetterQueueService {

>   @(restrict: [{ grant: '*', to: 'internal-user' }])

Reviewer: ```suggestion
  @requires('internal-user')
```

>   entity DeadOutboxEntries as projection on cds.outbox.Messages

Reviewer: ```suggestion
  entity DeadOutboxMessages as projection on cds.outbox.Messages
```

>   entity DeadOutboxEntries as projection on cds.outbox.Messages

Reviewer: Should we exclude properties from the projection?

Author: @BraunMatthias In my opinion we should include all properties when returning the entries for the sake of completeness.

    actions {
      action reactivate();
      action delete();
    };

}
```

:::

The `OutboxDeadLetterQueueService` provides an entity `DeadOutboxEntries` which is a projection on the outbox table `cds.outbox.Messages` that has two bound actions:

- `reactivate()` sets the number of attempts to `0` such that the outbox entry is going to be processed again.
- `delete()` deletes the outbox entry from the database.

Filters can be applied as for any other CDS defined entity, e.g. to filter for a specific outbox where the outbox name is stored in the field `target` of the entity `cds.outbox.Messages`.

::: warning

It is crucial to make the entity `DeadOutboxEntries` accessible for internal users only as it contains sensitive data that could be exploited for malicious purposes if unauthorized changes are performed.

[Learn more about pseudo roles](../guides/security/authorization#pseudo-roles){.learn-more}

:::

> To ensure that only dead outbox entries are returned when reading `DeadOutboxEntries`, an `@After` handler needs to be developped that filters for dead entries. This filtering can't be done on the database since the maximum number of attempts is only available from the CDS properties. The following code provides the handler for the `DeadLetterQueueService` and the `@After` read handler that filters for the dead outbox entries:

Reviewer: ```suggestion
To ensure that only dead outbox entries are returned when reading `DeadOutboxEntries`, an `@After` handler needs to be added which filters the entries. This filtering can't be done on the database since the maximum number of attempts is only available from the CDS properties. The following code provides the handler for the `DeadLetterQueueService` and the `@After-READ` handler that filters for the dead outbox entries:
```

> To ensure that only dead outbox entries are returned when reading `DeadOutboxEntries`, an `@After` handler needs to be developped that filters for dead entries. This filtering can't be done on the database since the maximum number of attempts is only available from the CDS properties. The following code provides the handler for the `DeadLetterQueueService` and the `@After` read handler that filters for the dead outbox entries:

Reviewer 2: Out of curiosity: this effectively means that if I want to generally increase the number of retries, older entries which have reached their max attempts in the past would be retried automatically. If the dead letter queue has a big size, this will lead to unintended load on the system.

Author: You are right. This behaviour existed from the very beginning of the outbox.

Author: @StefanHenke I adding this fact as a warning.

```java
@Component
@ServiceName(OutboxDeadLetterQueueService_.CDS_NAME)
public class DeadOutboxEntriesHandler implements EventHandler {

> 	@After(event = CqnService.EVENT_READ, entity = DeadOutboxEntries_.CDS_NAME)

Reviewer: ```suggestion
	@After
```

	public void filterDeadEntries(CdsReadEventContext context) {
		CdsProperties.Outbox outboxConfigs = context.getCdsRuntime().getEnvironment().getCdsProperties().getOutbox();
		List<DeadOutboxEntries> deadEntries = context
				.getResult()
				.listOf(DeadOutboxEntries.class)
				.stream()
				.filter(entry -> entry.getAttempts() >= outboxConfigs.getService(entry.getTarget()).getMaxAttempts())
				.toList();

		context.setResult(deadEntries);
	}
}
```

[Learn more about event handlers.](./event-handlers/){.learn-more}

Next, to implement the functionality for the bound actions (`reactivate` and `delete`) defined for the `DeadOutboxEntries` entity, you need to create the corresponding handlers:

```java
@Autowired
private OutboxDeadLetterQueueService outboxDeadLetterQueueService;

@On
public void reactivateOutboxEntry(DeadOutboxEntriesReactivateContext context) {
  CqnAnalyzer analyzer = CqnAnalyzer.create(context.getModel());

>   AnalysisResult analysisResult = analyzer.analyze(context.getCqn());

Reviewer: ```suggestion
  AnalysisResult analysisResult = CqnAnalyzer.create(context.getModel()).analyze(context.getCqn());
```

  Map<String, Object> key = analysisResult.rootKeys();

>   this.outboxDeadLetterQueueService.run(Update.entity(DeadOutboxEntries_.class).entry(key).data(Map.of(Messages.ATTEMPTS, 0)));

Reviewer: Better work typesafe with the DeadOutBoxEntries-POJO

  context.setCompleted();
}

@On
public void deleteOutboxEntry(DeadOutboxEntriesDeleteContext context) {
  CqnAnalyzer analyzer = CqnAnalyzer.create(context.getModel());
  AnalysisResult analysisResult = analyzer.analyze(context.getCqn());

>   Map<String, Object> key = analysisResult.rootKeys();

Reviewer: ```suggestion
  Map<String, Object> key = CqnAnalyzer.create(context.getModel()).analyze(context.getCqn()).rootKeys();
```

| `com.sap.cds.outbox.coldEntries`           | Number of entries that could not be delivered after repeated attempts and will not be retried anymore. | Gauge    |
| `com.sap.cds.outbox.remainingEntries`      | Number of entries which are pending for delivery.                                                      | Gauge    |
| `com.sap.cds.outbox.maxStorageTimeSeconds` | Maximum time in seconds an entry was residing in the outbox.                                           | Gauge    |
| `com.sap.cds.outbox.medStorageTimeSeconds` | Median time in seconds of an entry stored in the outbox."                                              | Gauge    |
| `com.sap.cds.outbox.minStorageTimeSeconds` | Minimal time in seconds an entry was stored in the outbox.                                             | Gauge    |
| `com.sap.cds.outbox.incomingMessages`      | Number of incoming messages of the outbox.                                                             | Counter  |
| `com.sap.cds.outbox.outgoingMessages`      | Number of outgoing messages of the outbox.                                                             | Counter  |

> The KPIs are logged per microservice instance (in case of horizontal scaling), outbox and tenant.

Reviewer 2: When we tried to build a dashboard for Cloud Logging Service we observed a few things which are difficult to handle. These are also worth to be mentioned here:
- the metrics are only send while the `PartitionCollector` is active. So, if it is not running, no metrics are reported to CLS which will interpret the values as zero. This is especially interesting for absolute values for `coldEntries` or `remainingEntries`
- If multiple microservice instances are sending the same metrics, there might be slightly different values available for the same time interval (e.g. minutes) as the instance will not meter at the exact same time within this time interval

[Learn more about `EventContext.proceed()`.](./event-handlers/#proceed-on){.learn-more}

## Outbox Dead Letter Queue

The transactional outbox tries to process each entry a specific number of times. The number of attempts is configurable per outbox by setting the configuration `cds.outbox.services.<key>.maxAttempts`.

[Learn more about CDS Properties.](./developing-applications/properties){.learn-more}

> Once the maximum number of attempts is exceeded, the corresponding entry is not touched anymore and hence it can be regarded as dead. Dead outbox entries are not deleted automatically, they remain in the database forever. By defining a CDS service the dead entries can be managed. The example below provides a step-by-step guide to develop a Dead Letter Queue for the transactional outbox.

Reviewer: ```suggestion
Once the maximum number of attempts is exceeded, the corresponding entry is not touched anymore and hence it can be regarded as dead. Dead outbox entries are not deleted automatically, they remain in the database and it is up to the application to take care of the entries. By defining a CDS service, the dead entries can be managed conveniently. The example below provides a step-by-step guide to develop a Dead Letter Queue for the transactional outbox.
```

::: warning

It is possible to increase the value of the configuration `cds.outbox.services.<key>.maxAttempts` in between of deployments. Older entries which have reached their max attempts in the past would be retried automatically after deployment of the new microservice version. If the dead letter queue has a big size, this will lead to unintended load on the system.

:::

In the first step the service needs to be defined in CDS:

::: code-group

```cds [srv/outbox-dead-letter-queue-service.cds]
using from '@sap/cds/srv/outbox';

service OutboxDeadLetterQueueService {

  @requires: 'internal-user'

>   entity DeadOutboxMessages as projection on cds.outbox.Messages

Reviewer: I'd prefer to have the projection entity `@readonly`

    actions {

>       action reactivate();

Reviewer: `reset()` and `remove() `?

Author: @BraunMatthias My proposal: `revive()` (as it's dead) and keep delete as that is what its doing.

- `delete()` deletes the outbox entry from the database.

Filters can be applied as for any other CDS defined entity, e.g. to filter for a specific outbox where the outbox name is stored in the field `target` of the entity `cds.outbox.Messages`.

::: warning

It is crucial to make the entity `DeadOutboxMessages` accessible for internal users only as it contains sensitive data that could be exploited for malicious purposes if unauthorized changes are performed.

[Learn more about pseudo roles](../guides/security/authorization#pseudo-roles){.learn-more}

:::

To ensure that only dead outbox entries are returned when reading `DeadOutboxMessages`, an `@After` handler needs to be added which filters the entries. This filtering can't be done on the database since the maximum number of attempts is only available from the CDS properties. The following code provides the handler for the `DeadLetterQueueService` and the `@After-READ` handler that filters for the dead outbox entries:

```java
@Component
@ServiceName(OutboxDeadLetterQueueService_.CDS_NAME)
public class DeadOutboxMessagesHandler implements EventHandler {

> 	@After

Reviewer: surprised that we don't need to specify the target entity `DeadOutboxMessages._CDS_NAME` here

Author: @BraunMatthias You proposed the change 🤣

Author: @BraunMatthias And it is working; nevertheless I'm going to add the event again

  deadOutboxMessage.setAttempts(0);

  this.outboxDeadLetterQueueService.run(Update.entity(DeadOutboxMessages_.class).entry(key).data(deadOutboxMessage));
  context.setCompleted();
}

@On
public void deleteOutboxEntry(DeadOutboxMessagesDeleteContext context) {
  AnalysisResult analysisResult = CqnAnalyzer.create(context.getModel()).analyze(context.getCqn());
  Map<String, Object> key = CqnAnalyzer.create(context.getModel()).analyze(context.getCqn()).rootKeys();

  this.outboxDeadLetterQueueService.run(Delete.from(DeadOutboxMessages_.class).byId(key.get(DeadOutboxMessages.ID)));
  context.setCompleted();
}
```

The injected `OutboxDeadLetterQueueService` instance is used to perform the operations on the entity. Both handlers first retrieve the ID of the entry and then they perform the corresponding operation on the database.

[Learn more about CQL statement inspection.](./working-with-cql/query-introspection#cqnanalyzer){.learn-more}

::: tip Use paging logic
Avoid to read all entries of the `cds.outbox.Messages` table at once, as the size of an entry is unpredictable
and depends on the size of the payload. Prefer paging logic instead.
:::

> ## Observability using Open Telemetry

Reviewer: Pls add a hint to 'cds add telemetry --...' command here to simplify configuration

Reviewer 2: Shouldn´t we better add the hint to the general OpenTelemetry docu which is linked here already: "./operating-applications/observability#open-telemetry"

I think we haven´t done that so far

Author: @StefanHenke @BraunMatthias I agree with Stefan as there is already a description how to enable OpenTelemetry. I would add a tip there that `cds add` can also be used.

:::

### Define the Service

::: code-group

```cds [srv/outbox-dead-letter-queue-service.cds]
using from '@sap/cds/srv/outbox';

@requires: 'internal-user'
service OutboxDeadLetterQueueService {

  @readonly
  entity DeadOutboxMessages as projection on cds.outbox.Messages
    actions {
      action revive();
      action delete();
    };

}
```

:::

The `OutboxDeadLetterQueueService` provides an entity `DeadOutboxMessages` which is a projection on the outbox table `cds.outbox.Messages` that has two bound actions:

- `revive()` sets the number of attempts to `0` such that the outbox entry is going to be processed again.
- `delete()` deletes the outbox entry from the database.

Filters can be applied as for any other CDS defined entity, e.g. to filter for a specific outbox where the outbox name is stored in the field `target` of the entity `cds.outbox.Messages`.

> ::: warning `DeadOutboxMessages` for internal users only

Reviewer 3: In the model, the whole service is internal. Would it be enough to just protect the entity?

service OutboxDeadLetterQueueService {

>   @requires: 'internal-user'

Reviewer: This should be on service level
